packages/@aws-cdk/aws-iot-actions/lib/s3-action.ts

```diff
@@ -0,0 +1,77 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as s3 from '@aws-cdk/aws-s3';
+import { kebab as toKebabCase } from 'case';
+import { singletonActionRole } from './private/role';
+
+/**
+ * Configuration properties of an action for s3.
+ */
+export interface S3ActionProps {
+  /**
+   * The Amazon S3 canned ACL that controls access to the object identified by the object key.
+   * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
+   *
+   * @default None
+   */
+  readonly cannedAcl?: s3.BucketAccessControl;
+
+  /**
+   * The path to the file where the data is written.
+   *
+   * Supports substitution templates.
+   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-substitution-templates.html
+   *
+   * @default '${topic()}/${timestamp()}'
+   */
+  readonly key?: string;
+
+  /**
+   * The IAM role that allows access to the S3.
+   *
+   * @default a new role will be created
+   */
+  readonly role?: iam.IRole;
+}
+
+/**
+ * The action to write the data from an MQTT message to an Amazon S3 bucket.
+ */
+export class S3Action implements iot.IAction {
+  private readonly cannedAcl?: string;
+  private readonly key?: string;
+  private readonly role?: iam.IRole;
+
+  /**
+   * @param bucket The Amazon S3 bucket to which to write data.
+   * @param props Optional properties to not use default
+   */
+  constructor(private readonly bucket: s3.IBucket, props: S3ActionProps = {}) {
+    this.cannedAcl = props.cannedAcl;
+    this.key = props.key;
+    this.role = props.role;
+  }
+
+  bind(rule: iot.ITopicRule): iot.ActionConfig {
+    const role = this.role ?? singletonActionRole(rule);
+    role.addToPrincipalPolicy(this.putEventStatement(this.bucket));
+
+    return {
+      configuration: {
+        s3: {
+          bucketName: this.bucket.bucketName,
+          cannedAcl: this.cannedAcl && toKebabCase(this.cannedAcl.toString()),
+          key: this.key ?? '${topic()}/${timestamp()}',
+          roleArn: role.roleArn,
+        },
+      },
+    };
+  }
+
+  private putEventStatement(bucket: s3.IBucket) {
+    return new iam.PolicyStatement({
+      actions: ['s3:PutObject'],
```

Comment:
Hmmm. Are you 100% sure this is the only needed permissions? Why not rely on `this.bucket.grantRead()` here, instead of creating a separate statement?

Reply:
Yup, needed is only `s3:PutObject` as [this docs](https://docs.aws.amazon.com/iot/latest/developerguide/s3-rule-action.html#s3-rule-action-requirements). And I confirmed that it works.

From the point of view of maintainability and understandability and simplicity of code, using `this.bucket.grantWrite()` is good, I think.
But now, I choice to keep minimal permissions. 

Witch is better do you think? If it is needed, I will refactor about also previous comments [this](https://github.com/aws/aws-cdk/pull/17110#discussion_r736849962) and [this](https://github.com/aws/aws-cdk/pull/17228#discussion_r740301268) in another PR!

Reply:
Ok, let's keep it only for `'s3:PutObject'`. You've already done the work, so we won't be saving anything anyway 😉.